partiql-eval/src/eval.rs

```diff
@@ -307,45 +331,52 @@ pub enum EvalBinop {
     Exp,
 }
 
-impl EvalExpr for EvalBinopExpr {
+impl EvalExpr for EvalBinOpExpr {
```

Comment:
Propagation and short-circuit logic non-localized here.

If we want `AND`/`OR` to short-circuit on booleans, and all other ops to short-circuit on null-propagation, something like the following may clean this up a bit:

```rust


impl EvalExpr for EvalBinopExpr {
    fn evaluate(&self, bindings: &Tuple, ctx: &dyn EvalContext) -> Value {
        #[inline]
        fn short_circuit(op: &EvalBinop, value: &Value) -> Option<Value> {
            match (op, value) {
                (EvalBinop::And, Value::Boolean(false)) => Some(false.into()),
                (EvalBinop::Or, Value::Boolean(true)) => Some(true.into()),
                (_, Value::Missing) => Some(Value::Missing),
                (_, Value::Null) => Some(Value::Null),
                _ => None,
            }
        }

        let lhs = self.lhs.evaluate(bindings, ctx);
        if let Some(propagate) = short_circuit(&self.op, &lhs) {
            return propagate;
        }

        let rhs = self.rhs.evaluate(bindings, ctx);
        match self.op {
            EvalBinop::And => lhs.and(rhs),
            EvalBinOp::Or => lhs.or(rhs),
            EvalBinOp::Concat => ...
            EvalBinOp::Eq => Value::from(lhs == rhs),
            EvalBinOp::Neq => Value::from(lhs != rhs),
            EvalBinOp::Gt => Boolean(lhs > rhs),
            EvalBinOp::Gteq => Boolean(lhs >= rhs),
            EvalBinOp::Lt => Boolean(lhs < rhs),
            EvalBinOp::Lteq => Boolean(lhs <= rhs),
            EvalBinOp::Add => lhs + rhs,
            EvalBinOp::Sub => lhs - rhs,
            EvalBinOp::Mul => lhs * rhs,
            EvalBinOp::Div => lhs / rhs,
            EvalBinOp::Mod => lhs % rhs,
            EvalBinOp::Exp => todo!("Exponentiation"),
        }
    }
}

```

Reply:
Thanks for the refactoring suggestion. Think the approach you've outlined is cleaner. From our offline discussion, made some adjustments to also check the RHS for `MISSING` and `NULL` values for equality and comparison ops.